cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1732,44 +1752,43 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
 
     auto outcome = client_->HeadObject(req);
     if (outcome.IsSuccess()) {
-      *out = true;
-      return Status::OK();
+      return true;
     }
     if (!backend_) {
       SaveBackend(outcome.GetError());
```

Comment:
Is `SaveBackend` an expensive operation?  If we move this check up above `if (backend_ && *backend_ == S3Backend::Minio)` it seems we could avoid potentially recursing by always getting it right the first time.  But I might be missing something.

Reply:
`SaveBackend` needs to inspect the headers from an error response, hence its signature.